### ISSUE TITLE
Support for quoted strings and string interpolation in the env_file format

### ISSUE THREAD

**Issue by Yogarine** [compose-spec/compose-spec#112]
**What is the problem you're trying to solve**
The current compose spec says:
> The value of VAL is used as a raw string and not modified at all. If the value is surrounded by quotes (as is often the case for shell variables), the quotes MUST be included in the value passed to containers created by the Compose implementation.

However, this is incompatible with most other common .env format specifications, which usually will use quotes to appropriately escape strings, and thus allow for multi-line strings.

The current version of docker-compose uses this more common .env format, which will not include surrounding quotes. Moreover, docker-compose allows for string interpolation of environment variables identical to the Compose spec Interpolation syntax, which is a very useful feature, and greatly cleans up docker-compose files by not having to resort to declaring environment variables in the `environment:` key just to be able to use interpolated environment variables or use multi-line strings.

However, if docker-compose where to be made compatible with the current Compose spec, it would mean removing this functionality.

**Describe the solution you'd like**
I'd like the Compose spec to be updated to follow the current functionality of docker-compose. e.g.:

"The value of VAL is a string that can optionally be surrounded by quotes. If the value is surrounded by quotes (as is often the case for shell variables), the quotes MUST NOT be included in the value passed to containers created by the Compose implementation.

The value of VAL MUST be interpolated using the Compose Interpolation syntax."

**Comment by EricHripko** [comment]
> However, this is incompatible with most other common .env format specifications, which usually will use quotes to appropriately escape strings, and thus allow for multi-line strings.

Could you please share what formats you're referring to? I am not aware of any spec for `.env` files, so behaviour here appears to be specific to libraries and/or tooling used.

This being a breaking change would also make it more difficult to introduce this functionality.

**Comment by ndeloof** [comment]
env file support by compose-spec is inherited by docker-compose's python codebase
so format is set by https://pypi.org/project/python-dotenv/

I'm not sure about quotes support there, but for compatibility with existing compose users whe should try to align. Also consider we should ignore any `export` prefix.

If there's any relevant dotEnv file format specification, we should just link ot it.

**Comment by Yogarine** [comment]
@EricHripko 
> Could you please share what formats you're referring to? I am not aware of any spec for `.env` files, so behaviour here appears to be specific to libraries and/or tooling used.

Apologies. I wrote "spec" but I guess I should've said "implementation". In this case the python-dotenv implementation seems to be the de-facto specification docker-compose uses.

> This being a breaking change would also make it more difficult to introduce this functionality.

It's a breaking change in the Compose spec, sure, but not in the current docker-compose implementation. Quoted strings and interpolation are already supported and as such docker-compose is currently diverging from the Compose spec (and their own documentation, for that matter). Considering the chance is very slim that docker-compose is going to switch their dotEnv parser from python-dotenv to a more specific and limited implementation, I recommend the Compose spec is updated to be brought in line with the current major implementation.


@ndeloof 
> env file support by compose-spec is inherited by docker-compose's python codebase
> so format is set by https://pypi.org/project/python-dotenv/
> 
> I'm not sure about quotes support there, but for compatibility with existing compose users whe should try to align. Also consider we should ignore any `export` prefix.

From that same page:

> .env looks like this:
> ```
> # a comment that will be ignored.
> REDIS_ADDRESS=localhost:6379
> MEANING_OF_LIFE=42
> MULTILINE_VAR="hello\nworld"
> ```

and:

> Ensure that variables are surrounded with {} like ${HOME} as bare variables such as $HOME are not expanded.
> ```
> CONFIG_PATH=${HOME}/.config/foo
> DOMAIN=example.org
> EMAIL=admin@${DOMAIN}
> DEBUG=${DEBUG:-false}
> ```

I really do like the idea of ignore `export` prefixes, that would simplify my workflow a lot.


> If there's any relevant dotEnv file format specification, we should just link ot it.

I don't believe there is an _official_ dotEnv file format specification. Like I mentioned above, it seems python-dotenv is currently the de-facto dotEnv spec as implemented here:
https://github.com/theskumar/python-dotenv/blob/master/src/dotenv/parser.py

**Comment by ndeloof** [comment]
Right, if we can't find a reasonable reference for dot env file format, we could include this in the compose-spec. imho this would make sense to create de dedicated page for it, so that it can be well defined without making the (already long) spec.md even longer

**Comment by EricHripko** [comment]
It appears that `compose-go` library currently rolls it's [own parser](https://github.com/compose-spec/compose-go/blob/3eee68c9fd46f0b8bcc594f582d09515f9966d5a/envfile/envfile.go) while `docker-compose` indeed uses [dotenv](https://github.com/docker/compose/blob/b01601a53cc8839afbe332f00d0d1b8165012e7b/compose/config/environment.py#L38). It does look like there's a need to define this behaviour better - otherwise, we risk different implementations having slightly different parsing rules.

**Comment by ndeloof** [comment]
`env_file` can be set with a `format: raw` sub-attribute if you want to disable compose-go "custom" dotenv parser, and get something closer to other libraries. dot.env file format isn't well defined anyway, so ...
